File: bin/carve/sheet.py
#!/usr/bin/env python3
"""工程の絵を 4 方向 × 3 コマの一覧に並べる (標準ライブラリのみ)。

    python3 bin/sheet.py 2_陰影

1 枚ずつ見ても方向のそろいは分からない。並べて初めて、背の高さ・接地・
色が方向で割れていないかが読める。
"""
import importlib.util, os, struct, sys, zlib

HERE = os.path.dirname(os.path.abspath(__file__))
_s = importlib.util.spec_from_file_location("render", os.path.join(HERE, "render.py"))
render = importlib.util.module_from_spec(_s); _s.loader.exec_module(render)

VIEWS = ("front", "east", "back", "west")
FRAMES = ("idle", "walk_0", "walk_1", "walk_2", "walk_3",
          "swing_0", "swing_1", "swing_2",
          "jump_0", "jump_1", "jump_2")
BACK = (30, 30, 34, 255)


def read_png(path):
    data = open(path, "rb").read()
    pos, w, h, idat = 8, None, None, b""
    while pos < len(data):
        ln = struct.unpack(">I", data[pos:pos + 4])[0]
        kind = data[pos + 4:pos + 8]
        if kind == b"IHDR":
            w, h = struct.unpack(">II", data[pos + 8:pos + 16])
        elif kind == b"IDAT":
            idat += data[pos + 8:pos + 8 + ln]
        pos += 12 + ln
    raw = zlib.decompress(idat)
    stride, rows, prev, i = w * 4, [], bytearray(w * 4), 0
    for _ in range(h):
        filt = raw[i]; i += 1
        line = bytearray(raw[i:i + stride]); i += stride
        for x in range(stride):
            a = line[x - 4] if x >= 4 else 0
            b = prev[x]
            c = prev[x - 4] if x >= 4 else 0
            if filt == 1: line[x] = (line[x] + a) & 255
            elif filt == 2: line[x] = (line[x] + b) & 255
            elif filt == 3: line[x] = (line[x] + (a + b) // 2) & 255
            elif filt == 4:
                p = a + b - c
                pa, pb, pc = abs(p - a), abs(p - b), abs(p - c)
                line[x] = (line[x] + (a if pa <= pb and pa <= pc
                                      else (b if pb <= pc else c))) & 255
        prev = line
        rows.append([tuple(line[x * 4:x * 4 + 4]) for x in range(w)])
    return w, h, rows


def main(argv):
    stage = argv[0] if argv else "2_陰影"
    root = os.path.dirname(HERE)
    out = []
    for body in sorted(os.listdir(os.path.join(root, "gallery"))):
        home = os.path.join(root, "gallery", body, stage)
        if not os.path.isdir(home):
            continue
        tiles = [read_png(os.path.join(home, f"{v}_{f}.png"))
                 for v in VIEWS for f in FRAMES]
        tw = max(t[0] for t in tiles); th = max(t[1] for t in tiles)
        cols = len(FRAMES) * len(VIEWS)
        canvas = [[BACK] * (tw * cols) for _ in range(th)]
        for i, (w, h, px) in enumerate(tiles):
            ox = i * tw + (tw - w) // 2
            oy = th - h
            for y in range(h):
                for x in range(w):
                    if px[y][x][3]:
                        canvas[oy + y][ox + x] = px[y][x]
        path = os.path.join(root, "gallery", f"_一覧_{body}_{stage}.png")
        with open(path, "wb") as handle:
            handle.write(render.png_of(tw * cols, th, canvas))
        out.append(os.path.relpath(path, root))
    print("\n".join(out))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: go/internal/carve/sheet.go
package carve

// 工程の絵を 4 方向 × コマの一覧に並べる。
//
//	fge-go carve-sheet 2_陰影

import (
	"bytes"
	"compress/zlib"
	"encoding/binary"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// SheetViews は一覧に並べる向き。
var SheetViews = []string{"front", "east", "back", "west"}

// SheetFrames は一覧に並べるコマ。
var SheetFrames = []string{"idle", "walk_0", "walk_1", "walk_2", "walk_3",
	"swing_0", "swing_1", "swing_2",
	"jump_0", "jump_1", "jump_2"}

// ReadPNGSheet は RGBA8 の PNG を読む。
//
// WhyNot: pxlib.ReadPNG に寄せないのは、こちらが 4 チャンネル決め打ちで
// 色票つきを読めないため。読める絵の範囲まで含めて写す。
func ReadPNGSheet(path string) (int, int, [][]RGBA, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return 0, 0, nil, err
	}
	pos, w, h := 8, 0, 0
	var idat []byte
	for pos+8 <= len(data) {
		ln := int(binary.BigEndian.Uint32(data[pos : pos+4]))
		kind := string(data[pos+4 : pos+8])
		switch kind {
		case "IHDR":
			w = int(binary.BigEndian.Uint32(data[pos+8 : pos+12]))
			h = int(binary.BigEndian.Uint32(data[pos+12 : pos+16]))
		case "IDAT":
			idat = append(idat, data[pos+8:pos+8+ln]...)
		}
		pos += 12 + ln
	}
	zr, err := zlib.NewReader(bytes.NewReader(idat))
	if err != nil {
		return 0, 0, nil, err
	}
	raw, err := io.ReadAll(zr)
	if err != nil {
		return 0, 0, nil, err
	}
	stride := w * 4
	rows := make([][]RGBA, 0, h)
	prev := make([]byte, stride)
	i := 0
	for y := 0; y < h; y++ {
		filt := raw[i]
		i++
		line := make([]byte, stride)
		copy(line, raw[i:i+stride])
		i += stride
		for x := 0; x < stride; x++ {
			var a, c int
			if x >= 4 {
				a, c = int(line[x-4]), int(prev[x-4])
			}
			b := int(prev[x])
			switch filt {
			case 1:
				line[x] = byte((int(line[x]) + a) & 255)
			case 2:
				line[x] = byte((int(line[x]) + b) & 255)
			case 3:
				line[x] = byte((int(line[x]) + (a+b)/2) & 255)
			case 4:
				p := a + b - c
				pa, pb, pc := absInt(p-a), absInt(p-b), absInt(p-c)
				near := c
				if pa <= pb && pa <= pc {
					near = a
				} else if pb <= pc {
					near = b
				}
				line[x] = byte((int(line[x]) + near) & 255)
			}
		}
		prev = line
		row := make([]RGBA, w)
		for x := 0; x < w; x++ {
			row[x] = RGBA{int(line[x*4]), int(line[x*4+1]), int(line[x*4+2]), int(line[x*4+3])}
		}
		rows = append(rows, row)
	}
	return w, h, rows, nil
}

func absInt(v int) int {
	if v < 0 {
		return -v
	}
	return v
}

// RunSheet は carve-sheet の入口。root は assets / gallery を置く根。
func RunSheet(out *strings.Builder, root string, args []string, rules *Rules) (int, error) {
	stage := "2_陰影"
	if len(args) > 0 {
		stage = args[0]
	}
	var made []string
	bodies, err := sortedDir(filepath.Join(root, "gallery"))
	if err != nil {
		return 1, err
	}
	for _, body := range bodies {
		home := filepath.Join(root, "gallery", body, stage)
		if !isDir(home) {
			continue
		}
		type tile struct {
			w, h int
			px   [][]RGBA
		}
		var tiles []tile
		for _, v := range SheetViews {
			for _, f := range SheetFrames {
				w, h, px, err := ReadPNGSheet(filepath.Join(home, fmt.Sprintf("%s_%s.png", v, f)))
				if err != nil {
					return 1, err
				}
				tiles = append(tiles, tile{w, h, px})
			}
		}
		tw, th := 0, 0
		for _, t := range tiles {
			if t.w > tw {
				tw = t.w
			}
			if t.h > th {
				th = t.h
			}
		}
		cols := len(SheetFrames) * len(SheetViews)
		canvas := make([][]RGBA, th)
		for y := range canvas {
			row := make([]RGBA, tw*cols)
			for x := range row {
				row[x] = rules.SheetBack
			}
			canvas[y] = row
		}
		for i, t := range tiles {
			ox := i*tw + (tw-t.w)/2
			oy := th - t.h
			for y := 0; y < t.h; y++ {
				for x := 0; x < t.w; x++ {
					if t.px[y][x][3] != 0 {
						canvas[oy+y][ox+x] = t.px[y][x]
					}
				}
			}
		}
		path := filepath.Join(root, "gallery", fmt.Sprintf("_一覧_%s_%s.png", body, stage))
		if err := os.WriteFile(path, PNGOf(tw*cols, th, canvas), 0o644); err != nil {
			return 1, err
		}
		rel, _ := filepath.Rel(root, path)
		made = append(made, rel)
	}
	fmt.Fprintln(out, strings.Join(made, "\n"))
	return 0, nil
}

func sortedDir(dir string) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	var names []string
	for _, e := range entries {
		names = append(names, e.Name())
	}
	sort.Strings(names)
	return names, nil
}

func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}
